File: typescript/base/src/main/java/lang/System.ts
import { FileOutputStream } from "../io/FileOutputStream";
import { PrintStream } from "../io/PrintStream";

export class System {
  private static readonly standardOutput = new FileOutputStream("/dev/stdout");
  private static readonly standardError = new FileOutputStream("/dev/stderr");

  public static readonly out = new PrintStream(System.standardOutput);
  public static readonly err = new PrintStream(System.standardError);

  public static exit(status: number): void {
    process.exit(status);
  }

  public static nanoTime(): number {
    if (process?.hrtime?.bigint) {
      return Number(process.hrtime.bigint());
    }
    return Date.now() * 1_000_000;
  }
}

File: cpp/base/src/main/java/lang/System.cpp
#include <cstdlib>
#include <chrono>

#include "java/lang/System.h"

namespace java {

java::FileOutputStream System::standardOutput("/dev/stdout");
java::FileOutputStream System::standardError("/dev/stderr");
java::PrintStream System::out(&System::standardOutput);
java::PrintStream System::err(&System::standardError);

[[noreturn]] void
System::exit(int status) {
    std::exit(status);
}

long long
System::nanoTime() {
    const auto now = std::chrono::steady_clock::now().time_since_epoch();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
}

}
